handleleave.go

```diff
@@ -98,3 +101,124 @@ func HandleMakeLeave(input HandleMakeLeaveInput) (*HandleMakeLeaveResponse, erro
 	}
 	return &makeLeaveResponse, nil
 }
+
+type CurrentStateQuerier interface {
+	CurrentStateEvent(ctx context.Context, roomID spec.RoomID, eventType string, stateKey string) (PDU, error)
+}
+
+// HandleSendLeave handles requests to `/send_leave
+// Returns the parsed event and an error.
+func HandleSendLeave(ctx context.Context,
+	requestContent []byte,
+	origin spec.ServerName,
+	roomVersion RoomVersion,
+	eventID, roomID string,
+	querier CurrentStateQuerier,
+	verifier JSONVerifier,
+) (PDU, error) {
```

Comment:
Sidenote: this function and others like it will likely need to be modified to support pseudo IDs, as in many cases we pull out the state key and check that it is a user ID, and then check that the domain is correct on it. I would probably structure this as:
 - static function `HandleSendLeave` as it is in this PR
 - Pull out the room version immediately.
 - Call `roomVer.HandleSendLeave(args...)` to let room versions decide how to implement this.
 - Most room versions will call a private function which does the below code, checking user IDs etc.
 - But pseudo IDs will call a custom function which omits these checks.

Whilst the caller _could_ call `roomVer.HandleSendLeave` this then creates bad symmetry, as there are cases where you don't know the room version at this point (e.g invites).